src/main/java/com/labsynch/labseer/service/ParentSwapStructuresServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+		}
+
+		// Check if there are existing errors or duplicates.
+		ParentValidationDTO validationDTO1;
+		ParentValidationDTO validationDTO2;
+		try {
+			validationDTO1 = parentService.validateUniqueParent(parent1);
+			validationDTO2 = parentService.validateUniqueParent(parent2);
+		} catch (Exception e) {
+			logger.error("Caught error while trying to validate parent", e);
+			return false;
+		}
+		if (validationDTO1.hasErrors()) {
+			logger.error(String.format("%s has existing errors: %s", corpName1, validationDTO1.getErrors()));
+			return false;
+		} else if (validationDTO2.hasErrors()) {
+			logger.error(String.format("%s has existing errors: %s", corpName2, validationDTO2.getErrors()));
+			return false;
+		} else if (!validationDTO1.isParentUnique() || !validationDTO2.isParentUnique()) {
+			logger.error(String.format("%s or %s have existing duplicates.", corpName1, corpName2));
+			return false;
+		}
+
+		// Check if duplicates get introduced on swapping the structures.
+		HashSet<String> dupeParents1;
+		HashSet<String> dupeParents2;
+		try {
+			dupeParents1 = getParents(parent2.getMolStructure(), parent1.getStereoCategory().getCode(), parent1.getStereoComment());
+			dupeParents2 = getParents(parent1.getMolStructure(), parent2.getStereoCategory().getCode(), parent2.getStereoComment());
+		} catch (Exception e) {
+			// In case of any exceptions, abort the swapping.
+			logger.error("Error finding parents to check for duplication", e);
+			return false;
+		}
+		// Duplicate with the items being swapped is okay.
+		dupeParents1.remove(corpName1);
+		dupeParents1.remove(corpName2);
+		dupeParents2.remove(corpName1);
+		dupeParents2.remove(corpName2);
+		if (dupeParents1.size() == 0 && dupeParents2.size() == 0) {
+			String mol = parent1.getMolStructure();
+			parent1.setMolStructure(parent2.getMolStructure());
+			parent2.setMolStructure(mol);
+			logger.info(String.format("Swapping corpName1=%s & corpName2=%s swap successful.", corpName1, corpName2));
+			return true;
+		} else {
+			logger.error(String.format("Swapping corpName1=%s & corpName2=%s creates duplicates.", corpName1, corpName2));
+			return false;
+		}
+    }
+
+	private HashSet<String> getParents(String molStructure, String stereoCategory, String stereoComment) throws CmpdRegMolFormatException {
+		ArrayList<Parent> matchingParents = new ArrayList<Parent>();
+		int[] parentCdIds = chemStructureService.checkDupeMol(molStructure, StructureType.PARENT);
+		for (int parentCdId : parentCdIds) {
+			for (Parent parent : Parent.findParentsByCdId(parentCdId).getResultList()) {
+				String parentStereoCategory = parent.getStereoCategory().getCode();
+				if (!parentStereoCategory.equalsIgnoreCase(stereoCategory)) {
+					continue;
+				}
+				String parentStereoComment = parent.getStereoComment();
+				if (parentStereoComment == null && stereoComment == null) {
+					matchingParents.add(parent);
+				} else if (parentStereoComment == null || stereoComment == null) {
```

Comment:
I think there's a possible complication when one is null and the other is empty string. Users don't have a way to truly null out a stereo comment, so we should consider null and empty string as equal to each other.
We handle this case elsewhere, although not that elegantly: https://github.com/mcneilco/acas-roo-server/blob/release/2022.1.x/src/main/java/com/labsynch/labseer/service/BulkLoadServiceImpl.java#L806